src/utils/simplifier.js

```diff
@@ -38,15 +39,68 @@ function simplify(realm, value: Value): Value {
     default:
       return value;
   }
 }
 
+function simplifyEquality(realm: Realm, equality: AbstractValue): Value {
+  let op = equality.kind;
+  let [x, y] = equality.args;
+  if (x instanceof ConcreteValue) [x, y] = [y, x];
+  if (x instanceof AbstractValue && x.kind === "conditional" && (!y.mightNotBeUndefined() || !y.mightNotBeNull())) {
+    // try to simplify "(cond ? xx : xy) op undefined/null" to just "cond" or "!cond"
+    let [cond, xx, xy] = x.args;
+    invariant(cond instanceof AbstractValue); // otherwise the the conditional should not have been created
+    if (op === "===" || op === "!==") {
+      // if xx === undefined && xy !== undefined then cond <=> x === undefined
+      if (!y.mightNotBeUndefined() && !xx.mightNotBeUndefined() && !xy.mightBeUndefined())
+        return op === "===" ? makeBoolean(realm, cond) : negate(realm, cond);
+      // if xx !== undefined && xy === undefined then !cond <=> x === undefined
+      if (!y.mightNotBeUndefined() && !xx.mightBeUndefined() && !xy.mightNotBeUndefined())
+        return op === "===" ? negate(realm, cond) : makeBoolean(realm, cond);
+      // if xx === null && xy !== null then cond <=> x === null
+      if (!y.mightNotBeNull() && !xx.mightNotBeNull() && !xy.mightBeNull())
+        return op === "===" ? makeBoolean(realm, cond) : negate(realm, cond);
+      // if xx !== null && xy === null then !cond <=> x === null
+      if (!y.mightNotBeNull() && !xx.mightBeNull() && !xy.mightNotBeNull())
+        return op === "===" ? negate(realm, cond) : makeBoolean(realm, cond);
+    } else {
+      invariant(op === "==" || op === "!=");
+      // if xx cannot be undefined/null and xy is undefined/null then !cond <=> x == undefined/null
+      if (!xx.mightBeUndefined() && !xx.mightBeNull() && (!xy.mightNotBeUndefined() || !xy.mightNotBeNull()))
+        return op === "==" ? negate(realm, cond) : makeBoolean(realm, cond);
+      // if xx is undefined/null and xy cannot be undefined/null then cond <=> x == undefined/null
+      if ((!xx.mightNotBeUndefined() || !xx.mightNotBeNull()) && !xy.mightBeUndefined() && !xy.mightBeNull())
+        return op === "==" ? makeBoolean(realm, cond) : negate(realm, cond);
+    }
+  }
+  return equality;
+}
+
+function makeBoolean(realm: Realm, value: Value): Value {
+  if (value.getType() === BooleanValue) return value;
+  if (value instanceof ConcreteValue) return new BooleanValue(realm, ToBoolean(realm, value));
+  invariant(value instanceof AbstractValue);
+  let v = AbstractValue.createFromUnaryOp(realm, "!", value, true, value.expressionLocation);
+  return AbstractValue.createFromUnaryOp(realm, "!", v, true, value.expressionLocation);
+}
+
 function negate(realm: Realm, value: Value): Value {
   if (value instanceof ConcreteValue) return ValuesDomain.computeUnary(realm, "!", value);
   invariant(value instanceof AbstractValue);
-  if (value.kind === "!") return value.args[0];
+  if (value.kind === "!") return makeBoolean(realm, value.args[0]);
```

Comment:
Great! Would be nice if we had a better framework to state assumptions/guarantees. Right now, I had to chase through all code paths to verify the intended boolean-ness of the result. This doesn't scale when building more complex simplification rules.

Reply:
There aren't any great ways of doing this that I can think of, but I'll get back to this when unifying all of the simplification code.